theano/gpuarray/c_code/spatialtf_sampler.c

```diff
@@ -0,0 +1,198 @@
+    void * alpha_p;
+    void * beta_p;
+    float af = alpha, bf = beta;
+    spatialtf_context_t spatialtf_ctx;
+    cudnnDataType_t dt;
+    // Number of color channels (feature maps) is the innermost dimension
+    cudnnTensorFormat_t tf = CUDNN_TENSOR_NHWC;
+    cudnnStatus_t err = CUDNN_STATUS_SUCCESS;
+
+    // Obtain grid dimensions
+    npy_int * dimensions_data = (npy_int *)PyArray_DATA( grid_dimensions );
+    const int num_images = dimensions_data[0];
+    const int height = dimensions_data[1];
+    const int width = dimensions_data[2];
+    const int num_channels = dimensions_data[3];
+
+    switch (grid->ga.typecode)
+    {
+    case GA_DOUBLE:
+        alpha_p = (void *)&alpha;
+        beta_p = (void *)&beta;
+        dt = CUDNN_DATA_DOUBLE;
+        break;
+    case GA_FLOAT:
+        alpha_p = (void *)&af;
+        beta_p = (void *)&bf;
+        dt = CUDNN_DATA_FLOAT;
+        break;
+    case GA_HALF:
+        alpha_p = (void *)&af;
+        beta_p = (void *)&bf;
+        dt = CUDNN_DATA_HALF;
+        break;
+    default:
+        PyErr_SetString( PyExc_TypeError,
+                         "Unsupported type in spatial transformer sampler" );
+        return -1;
+    }
+
+    spatialtf_context_init( &spatialtf_ctx );
+
+    cuda_enter( gpu_ctx->ctx );
+
+    err = cudnnCreateTensorDescriptor( &(spatialtf_ctx.xdesc) );
+
+    if ( err != CUDNN_STATUS_SUCCESS )
+    {
+        spatialtf_context_destroy( &spatialtf_ctx );
+        cuda_exit( gpu_ctx->ctx );
+
+        PyErr_Format( PyExc_RuntimeError,
+                      "Could not create xdesc: %s",
+                      cudnnGetErrorString(err) );
+        return -1;
+    }
+
+    err = cudnnSetTensor4dDescriptor( spatialtf_ctx.xdesc, tf, dt, num_images,
+        num_channels, height, width );
+
+    if ( err != CUDNN_STATUS_SUCCESS )
+    {
+        spatialtf_context_destroy( &spatialtf_ctx );
+        cuda_exit( gpu_ctx->ctx );
+
+        PyErr_Format( PyExc_RuntimeError,
+                      "Could not initialize xdesc: %s",
+                      cudnnGetErrorString(err) );
+        return -1;
+    }
+
+    err = cudnnCreateTensorDescriptor( &(spatialtf_ctx.ydesc) );
+
+    if ( err != CUDNN_STATUS_SUCCESS )
+    {
+        spatialtf_context_destroy( &spatialtf_ctx );
+        cuda_exit( gpu_ctx->ctx );
+
+        PyErr_Format( PyExc_RuntimeError,
+                      "Could not create xdesc: %s",
+                      cudnnGetErrorString(err) );
+        return -1;
+    }
+
+    err = cudnnSetTensor4dDescriptor( spatialtf_ctx.ydesc, tf, dt, num_images,
+        num_channels, height, width );
+
+    if ( err != CUDNN_STATUS_SUCCESS )
+    {
+        spatialtf_context_destroy( &spatialtf_ctx );
+        cuda_exit( gpu_ctx->ctx );
+
+        PyErr_Format( PyExc_RuntimeError,
+                      "Could not initialize ydesc: %s",
+                      cudnnGetErrorString(err) );
+        return -1;
+    }
+
+    if ( NULL == *output )
```

Comment:
I mean, like any other Theano output variable, you should check if either it is null or if it is allocated with bad shapes, and reallocate if so (as you do with `grids` in another C code).